plugins/usbdmx/AsyncUsbTransceiverBase.cpp

```diff
@@ -67,62 +101,69 @@ bool AsyncUsbTransceiverBase::Init() {
 }
 
 void AsyncUsbTransceiverBase::CancelTransfer() {
-  if (!m_transfer) {
-    return;
+  m_suppress_continuation = true;
+
+  inflight_set::iterator i;
+  for (i = m_inflight.begin(); i != m_inflight.end(); ++i) {
+    int ret = m_adaptor->CancelTransfer(*i);
+    if (ret != 0 && ret != LIBUSB_ERROR_NOT_FOUND) {
+      OLA_WARN << "libusb_cancel_transfer returned "
+               << m_adaptor->ErrorCodeToString(ret);
+    }
   }
 
-  bool canceled = false;
-  while (1) {
-    ola::thread::MutexLocker locker(&m_mutex);
-    if (m_transfer_state == IDLE || m_transfer_state == DISCONNECTED) {
-      break;
-    }
-    if (!canceled) {
-      m_suppress_continuation = true;
-      if (m_adaptor->CancelTransfer(m_transfer) == 0) {
-        canceled = true;
-      } else {
-        break;
-      }
-    }
+  while (!m_inflight.empty()) {
+    /* Wait for cancelled transfers to complete. Not waiting for the
+     * callbacks to happen is undefined behaviour according to libusb
+     * docs. */
+    m_cond.Wait(&m_mutex);
   }
+}
+
+struct libusb_transfer *AsyncUsbTransceiverBase::CurrentTransfer() {
+  if (m_idle.empty())
+    throw std::out_of_range("AsyncUsbTransceiverBase::m_idle is empty");
```

Comment:
We've not got any exception throwing in the code so far, returning NULL and then checking for that elsewhere would match the behaviour of the rest of the code better, possibly combined with a local log message. Or switching to assert().